src/main/java/com/rackspace/ceres/app/services/MetricDeletionService.java

```diff
@@ -0,0 +1,294 @@
+      Instant timeSlot, String metricName, Flux<String> seriesSetHashedEntries) {
+    return seriesSetHashedEntries.flatMap(seriesSetHashEntry -> {
+      return Flux.fromIterable(granularities).flatMap(granularity -> {
+        return cqlTemplate
+            .execute(dataTablesStatements.downsampleDeleteWithSeriesSetHash(granularity.getWidth()),
+                tenant,
+                timeSlot, seriesSetHashEntry);
+      }).then(Mono.just(true));
+    }).then(Mono.just(true))
+        .flatMap(item -> {
+          if (item) {
+            return seriesSetHashedEntries.flatMap(seriesSetHashEntry -> {
+              return cqlTemplate
+                  .execute(dataTablesStatements.getRawDeleteWithSeriesSetHash(), tenant, timeSlot,
+                      seriesSetHashEntry);
+            }).then(Mono.just(true));
+          } else {
+            return Mono.just(false);
+          }
+        })
+        .flatMap(result -> {
+          if (result) {
+            return deleteMetricNamesByTenantAndMetricName(tenant, metricName);
+          } else {
+            return Mono.just(false);
+          }
+        })
+        .flatMap(result -> {
+          if (result) {
+            return seriesSetHashedEntries.flatMap(seriesSetHashEntry -> {
+              return deleteSeriesSetHashes(tenant,
+                  seriesSetHashEntry);
+            }).then(Mono.just(true));
+          } else {
+            return Mono.just(false);
+          }
+        })
+        .flatMap(result -> {
+          if (result) {
+            return seriesSetHashedEntries.flatMap(seriesSetHashEntry -> {
+              return removeEntryFromCache(tenant,
+                  seriesSetHashEntry);
+            }).then(Mono.just(true));
+          } else {
+            return Mono.just(false);
+          }
+        })
+        .flatMap(result -> {
+          if (result) {
+            return deleteSeriesSetsByTenantIdAndMetricName(tenant, metricName);
+          } else {
+            return Mono.just(false);
+          }
+        });
+  }
+
+  private Mono<Boolean> deleteMetricsByTenantId(List<Granularity> granularities, String tenant,
+      Instant timeSlot) {
+    Flux<Map<String, Object>> seriesSetHashedEntries = getSeriesSetHashFromRaw(tenant, timeSlot);
+    Flux<Map<String, Object>> metricNameEntries = getMetricNames(tenant);
+    return Flux.fromIterable(granularities)
+        .flatMap(
+            granularity -> {
+              return cqlTemplate
+                  .execute(dataTablesStatements.downsampleDelete(granularity.getWidth()), tenant,
+                      timeSlot);
+            }).then(Mono.just(true))
+        .flatMap(result -> {
+          if (result) {
+            return seriesSetHashedEntries.flatMap(seriesSetHashEntry -> {
+              return deleteSeriesSetHashes(tenant,
+                  seriesSetHashEntry.get("series_set_hash").toString());
+            }).then(Mono.just(true));
+          } else {
+            return Mono.just(false);
```

Comment:
Why are we returning false here and continuing the chain? 
Can we just error the mono here?
What happens if data is deleted from first chain and failed in second one? Do we need to add support for retry?

Reply:
yes, returning error will be good in case of any failure and adding retry will be a good add-on too.